src/prpy/tsr/tsr.py

```diff
@@ -52,36 +61,130 @@ def __init__(self, T0_w=None, Tw_e=None, Bw=None,
+        Tw = kin.pose_to_H(kin.pose_from_xyzypr(xyzypr))
         trans = numpy.dot(numpy.dot(self.T0_w, Tw), self.Tw_e)
         return trans
 
+    def to_xyzrpy(self, trans):
+        """
+        Converts an end-effector transform to xyzrpy values
+        @param  trans  4x4 transform
+        @return xyzrpy 6x1 vector of Bw values
+        """
+        Tw = numpy.dot(numpy.dot(numpy.linalg.inv(self.T0_w), trans),
+                       numpy.linalg.inv(self.Tw_e))
+        pose = kin.pose_from_H(Tw)
+        ypr = kin.quat_to_ypr(pose[3:7])
+        xyzrpy = [pose[0], pose[1], pose[2],
+                  ypr[2], ypr[1], ypr[0]]
+        return xyzrpy
+
+    def is_valid(self, xyzrpy, ignoreNAN=False):
+        """
+        Checks if a xyzrpy is a valid sample from the TSR.
+        Two main issues: dealing with roundoff issues for zero bounds and
+        Wraparound for rpy.
+        @param xyzrpy 6x1 vector of Bw values
+        @param ignoreNAN (optional, defaults to False) ignore NaN xyzrpy
+        @return True if valid and False if not
+        """
+        Bw_xyz = self.Bw[0:3, :]
+
+        xyzcheck = [((x + EPSILON) >= Bw_xyz[i, 0]) and
+                    ((x - EPSILON) <= Bw_xyz[i, 1])
+                    for i, x in enumerate(xyzrpy[0:3])]
+
+        # Unwrap all rotations to [-pi, pi]
+        Bw_rpy = (self.Bw[3:6, :] + pi) % (2*pi) - pi
+        rpy = numpy.add(xyzrpy[3:6], pi) % (2*pi) - pi
+        rpycheck = [((x + EPSILON) >= Bw_rpy[i, 0]) and
+                    ((x - EPSILON) <= Bw_rpy[i, 1])
+                    for i, x in enumerate(rpy)]
+
+        check = numpy.hstack((xyzcheck, rpycheck))
+
+        # Strip out Nans if needed
+        if ignoreNAN:
+            check = check[~numpy.isnan(xyzrpy)]
+
+        return all(check)
+
+    def contains(self, trans):
+        """
+        Checks if the TSR contains the transform
+        @param  trans 4x4 transform
+        @return       True if inside and False if not
+        """
+        xyzrpy = self.to_xyzrpy(trans)
+        return self.is_valid(xyzrpy)
+
+    def distance(self, trans):
+        """
+        Computes the Geodesic Distance from the TSR to a transform
+        @param trans 4x4 transform
+        @return dist Geodesic distance to TSR
+        @return bwopt Closest Bw value to trans
+        """
+        if self.contains(trans):
+            return 0., self.to_xyzrpy(trans)
+
+        import scipy.optimize
+
+        def objective(bw):
+            bwtrans = self.to_transform(bw)
+            return prpy.util.GeodesicDistance(bwtrans, trans)
+
+        bwinit = (self.Bw[:, 0] + self.Bw[:, 1])/2
+        bwbounds = [(self.Bw[i, 0], self.Bw[i, 1]) for i in range(6)]
+
+        bwopt, dist, info = scipy.optimize.fmin_l_bfgs_b(
+                                objective, bwinit, fprime=None,
+                                args=(),
+                                bounds=bwbounds, approx_grad=True)
+        return dist, bwopt
```

Comment:
I suppose if we were to use this often, we could at least compute the gradient analytically, right?